Enable LTO in Rust release profile

ruokun-niu opened the pull request:
# Description

Enabling Link Time Optimization (LTO) will reduce the binary size. By setting `lto` to be `true` in the release profile, LTO will only be enabled in our release build, as we use the command `cargo install` in our dockerfiles. When we do a local `cargo build`, LTO will not be enabled. To enable LTO in `cargo build`, run with the `--release` flag

## Type of change

<!--

Please select **one** of the following options that describes your change and delete the others. Clearly identifying the type of change you are making will help us review your PR faster, and is used in authoring release notes.

If you are making a bug fix or functionality change to Drasi and do not have an associated issue link please create one now. 

-->

- This pull request is a minor refactor, code cleanup, test improvement, or other maintenance task and doesn't change the functionality of Drasi (issue link optional).

<!--

Please update the following to link the associated issue. This is required for some kinds of changes (see above).

-->

Fixes: [#97](https://github.com/drasi-project/drasi-platform/issues/97)

danielgerlag commented:
What about the development workflow where you need to build the docker image and load it to kind for local testing?
I think we need a flag on the make command?

ruokun-niu commented:
> What about the development workflow where you need to build the docker image and load it to kind for local testing? I think we need a flag on the make command?

Our makefile uses docker to build the image, which should use `cargo install` as well. When we run `cargo install`, the release profile will be used

danielgerlag commented:
> > What about the development workflow where you need to build the docker image and load it to kind for local testing? I think we need a flag on the make command?
> 
> Our makefile uses docker to build the image, which should use `cargo install` as well. When we run `cargo install`, the release profile will be used

Can we have a flag or ENV variable to control this, so that we can build the docker images without LTO by default?

ruokun-niu commented:
> > > What about the development workflow where you need to build the docker image and load it to kind for local testing? I think we need a flag on the make command?
> > 
> > 
> > Our makefile uses docker to build the image, which should use `cargo install` as well. When we run `cargo install`, the release profile will be used
> 
> Can we have a flag or ENV variable to control this, so that we can build the docker images without LTO by default?

I'll look into it

ruokun-niu commented:
> > > What about the development workflow where you need to build the docker image and load it to kind for local testing? I think we need a flag on the make command?
> > 
> > 
> > Our makefile uses docker to build the image, which should use `cargo install` as well. When we run `cargo install`, the release profile will be used
> 
> Can we have a flag or ENV variable to control this, so that we can build the docker images without LTO by default?

I have updated the makefiles and added the following section to the `contributing-code-building` readme:


By default, all our Rust-based images use the `release` profile in builds, which sets Link-time Optimization (LTO) to true. For more information on Rust profiles and the use of LTO in Rust projects, see [Profiles](https://doc.rust-lang.org/cargo/reference/profiles.html#lto). If you wish to use the default LTO setting, you can use the `CARGO_DEBUG` parameter as follows:
```sh
make docker-build CARGO_DEBUG=1
```

danielgerlag commented:
> > > > What about the development workflow where you need to build the docker image and load it to kind for local testing? I think we need a flag on the make command?
> > > 
> > > 
> > > Our makefile uses docker to build the image, which should use `cargo install` as well. When we run `cargo install`, the release profile will be used
> > 
> > 
> > Can we have a flag or ENV variable to control this, so that we can build the docker images without LTO by default?
> 
> I have updated the makefiles and added the following section to the `contributing-code-building` readme:
> 
> By default, all our Rust-based images use the `release` profile in builds, which sets Link-time Optimization (LTO) to true. For more information on Rust profiles and the use of LTO in Rust projects, see [Profiles](https://doc.rust-lang.org/cargo/reference/profiles.html#lto). If you wish to use the default LTO setting, you can use the `CARGO_DEBUG` parameter as follows:
> 
> ```shell
> make docker-build CARGO_DEBUG=1
> ```

Great, I think the default should be without LTO, so when you want to cut a release build it would be something like
```shell
make docker-build RELEASE=1
```

danielgerlag commented on control-planes/mgmt_api/Makefile:
```diff
@@ -4,11 +4,12 @@ CLUSTER_NAME ?= kind
 
 docker-build:
-	docker buildx build ../../ -f ./Dockerfile -t $(IMAGE_PREFIX)/api:$(DOCKER_TAG_VERSION) $(DOCKERX_OPTS)
+	docker buildx build ../../ --build-arg RELEASE=$(RELEASE) -f ./Dockerfile -t $(IMAGE_PREFIX)/api:$(DOCKER_TAG_VERSION) $(DOCKERX_OPTS)
```
We still need `$(DOCKERX_OPTS)`

danielgerlag commented on control-planes/mgmt_api/Dockerfile:
```diff
@@ -1,18 +1,20 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+ARG RELEASE=1
```
Let's make the default 0

danielgerlag replied:
Let's make the default 0

danielgerlag commented on control-planes/kubernetes_provider/Makefile:
```diff
@@ -4,11 +4,12 @@ CLUSTER_NAME ?= kind
 
 docker-build:
-	docker buildx build .. -f ./Dockerfile  -t $(IMAGE_PREFIX)/kubernetes-provider:$(DOCKER_TAG_VERSION) $(DOCKERX_OPTS)
+	docker buildx build .. --build-arg RELEASE=$(RELEASE) -f ./Dockerfile  -t $(IMAGE_PREFIX)/kubernetes-provider:$(DOCKER_TAG_VERSION) $(DOCKERX_OPTS)
```
could we just include this arg in the `$(DOCKERX_OPTS)` string?

danielgerlag replied:
Then you would not need to modify the make files at all, you could just pass it as a parameter

danielgerlag replied:
could we just include this arg in the $(DOCKERX_OPTS) string?
Then you would not need to modify the make files at all, you could just pass it as a parameter

danielgerlag commented on docs/contributing/contributing-code/contributing-code-building/README.md:
````diff
@@ -25,6 +25,12 @@ If you wish to override the default (latest) image tag, you can use the `DOCKER_
+By default, all our Rust-based images use the `release` profile in builds, which sets Link-time Optimization (LTO) to true. For more information on Rust profiles and the use of LTO in Rust projects, see [Profiles](https://doc.rust-lang.org/cargo/reference/profiles.html#lto). If you wish to use the default LTO setting, you can use the `CARGO_DEBUG` parameter as follows:
+
+```sh
+make docker-build CARGO_DEBUG=1
````
Is this still accurate?

ruokun-niu commented:
@drasi-project/drasi-engineering-team 
Hi team, I am closing this PR as I have moved the LTO changes to this PR (https://github.com/drasi-project/drasi-platform/pull/143). The reasoning behind this is that both PRs focused on supporting a debug build for our images (This PR only focuses on the Rust projects), and I believe it makes sense to combine them into one PR. For the Rust projects, I think LTO should be enabled by default and it should only be disabled in a debug build. Please let me know if there are any questions/concerns.
